packages/ui/src/pages/SecondarySignIn/index.tsx

```diff
@@ -1,44 +1,45 @@
+import { SignInIdentifier } from '@logto/schemas';
 import { useMemo } from 'react';
 import { useParams } from 'react-router-dom';
 
 import SecondaryPageWrapper from '@/components/SecondaryPageWrapper';
-import EmailSignIn from '@/containers/EmailForm/EmailSignIn';
-import { PhonePasswordless } from '@/containers/Passwordless';
+import { EmailSignIn } from '@/containers/EmailForm';
+import { SmsSignIn } from '@/containers/PhoneForm';
 import UsernameSignIn from '@/containers/UsernameSignIn';
 import { useSieMethods } from '@/hooks/use-sie';
 import ErrorPage from '@/pages/ErrorPage';
-import { UserFlow } from '@/types';
 
 type Props = {
   method?: string;
 };
 
 const SecondarySignIn = () => {
-  const { method = 'username' } = useParams<Props>();
+  const { method = '' } = useParams<Props>();
   const { signInMethods } = useSieMethods();
+  const signInMethod = signInMethods.find(({ identifier }) => identifier === method);
 
   const signInForm = useMemo(() => {
-    if (method === 'sms') {
+    if (signInMethod?.identifier === SignInIdentifier.Sms) {
       // eslint-disable-next-line jsx-a11y/no-autofocus
-      return <PhonePasswordless autoFocus type={UserFlow.signIn} />;
+      return <SmsSignIn autoFocus signInMethod={signInMethod} />;
     }
 
-    if (method === 'email') {
-      const signInMethod = signInMethods.find(({ identifier }) => identifier === method);
-
+    if (signInMethod?.identifier === SignInIdentifier.Email) {
       // eslint-disable-next-line jsx-a11y/no-autofocus
-      return signInMethod && <EmailSignIn autoFocus signInMethod={signInMethod} />;
+      return <EmailSignIn autoFocus signInMethod={signInMethod} />;
     }
 
-    // eslint-disable-next-line jsx-a11y/no-autofocus
-    return <UsernameSignIn autoFocus />;
-  }, [method, signInMethods]);
+    if (signInMethod?.identifier === SignInIdentifier.Username) {
+      // eslint-disable-next-line jsx-a11y/no-autofocus
+      return <UsernameSignIn autoFocus />;
+    }
+  }, [signInMethod]);
```

Comment:
I think couple months ago we agreed that we should not use `useMemo` for tsx elements. As `useMemo` is designed to get the complicated calculations cached, typically returning a javascript object out of it, instead of a conditional rendering tsx logic.

Simply write the conditional logics directly in the tsx part.

Reply:
Old logic updated